### PR TITLE
[flang]Fix linkage of globals

### PR DESCRIPTION
**Pull request by Leporacanthicus**
This fixes the issues with "undefined symbol" seen when optimising flang-compiled code.

This relies on the also included mlir fix cherry-picked from llvm-man.

This fixes https://github.com/flang-compiler/f18-llvm-project/issues/1095

### CODE REVIEW

**Review by clementval — APPROVED** [review]
Thanks for working on this. LGTM.

File: flang/lib/Lower/ConvertVariable.cpp [hunk 1]
```diff
@@ -1650,14 +1650,9 @@ void Fortran::lower::mapSymbolAttributes(
...
-  // even if it is unused and there is not init (other than undef).
-  // Emitting llvm :'@glob = global type undef' would also work, but mlir
-  // always insert "external" and removes the undef init when lowering a
-  // global without explicit linkage. This ends-up in llvm removing the
-  // symbol if unsued potentially creating linking issues. Hence the use
-  // of linkOnce.
-  auto linkOnce = converter.getFirOpBuilder().createLinkOnceLinkage();
+  // Use empty linkage for module variables, which makes them available
+  // for use in another unit.
+  auto linkage = converter.getFirOpBuilder().getStringAttr("");
```
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertVariable.cpp:1655]: Can we call this `externalLinkage` or something?

Also, I think you can abbreviate this line and make it bit more clear as well.
```c++
mlir::StringAttr externalLinkage;
```

**Review by schweitzpgi — COMMENTED** [review]
Thanks for the fix.

File: flang/lib/Lower/ConvertVariable.cpp (hunk 1 above)
Comment (Leporacanthicus) [reply flang/lib/Lower/ConvertVariable.cpp]: Done! Thanks for pointing this out!

**Review by schweitzpgi — APPROVED** [review]

**Pull request merged**